### ISSUE TITLE
"EMPTY" is a reserved word in CSharp target

### ISSUE THREAD

**Issue by kaby76** [antlr/antlr4#3199]
As the title says, EMPTY is a reserved symbol in grammars for the C# target (and likely in the alt tool/runtime Antlr4cs). That's because it's defined in ParserRuleContext [here](https://github.com/antlr/antlr4/blob/4b649103f31e0c0bc3bd0bb964efd0dea6b9f5a0/runtime/CSharp/src/ParserRuleContext.cs#L37). If you define EMPTY as a lexer symbol in your grammar, the compiler will give warnings. For further information, see [this issue](https://github.com/antlr/grammars-v4/issues/2225). The workaround is to just rename EMPTY to EMPTY_ in your grammar. (I use [trrename](https://github.com/kaby76/Domemtech.Trash/tree/main/trrename) to bulk rename such conflicts quickly.)

**Comment by KvanTTT** [comment]
Could be added to `badwords` list for C#.

**Comment by kaby76** [comment]
Ha! C# is the only language [without "bad words"](https://github.com/antlr/antlr4/blob/cb4e59d420d5749d457a37b2652732b5969e5d67/tool/src/org/antlr/v4/codegen/target/CSharpTarget.java#L56). Goodness, what's a language without of few colorful metaphors. (All the other targets have the API [Cpp](https://github.com/antlr/antlr4/blob/cb4e59d420d5749d457a37b2652732b5969e5d67/tool/src/org/antlr/v4/codegen/target/CppTarget.java#L59), [Dart](https://github.com/antlr/antlr4/blob/cb4e59d420d5749d457a37b2652732b5969e5d67/tool/src/org/antlr/v4/codegen/target/DartTarget.java#L65), [Go](https://github.com/antlr/antlr4/blob/cb4e59d420d5749d457a37b2652732b5969e5d67/tool/src/org/antlr/v4/codegen/target/GoTarget.java#L77), [JavaScript](https://github.com/antlr/antlr4/blob/cb4e59d420d5749d457a37b2652732b5969e5d67/tool/src/org/antlr/v4/codegen/target/JavaScriptTarget.java#L57), [Java](https://github.com/antlr/antlr4/blob/cb4e59d420d5749d457a37b2652732b5969e5d67/tool/src/org/antlr/v4/codegen/target/JavaTarget.java#L52), etc.)

**Comment by KvanTTT** [comment]
This is because C# allows [keyword escaping](https://stackoverflow.com/a/421263/1046374). But `EMPTY` seems like is not the case and I guess such a set should be added to C#.

**Comment by kaby76** [comment]
Goodness, why wasn't this mapping added for each of the other targets? This seems like an easy change and it would have prevented hundreds, if not thousands, of renames in grammars-v4 to remove symbol conflicts that I and others manual made over many months. (I implemented and used [trrename](https://github.com/kaby76/Domemtech.Trash/tree/main/trrename) a few months ago to change whole-scale the remaining grammars--half--in grammars-v4 to complete the renaming to eliminate the symbol conflicts.) Symbol rewrite in the generated code was [suggested in 2015](https://github.com/antlr/antlr4/issues/1070) but nothing came of it.

_stop, _start wouldn't have to be added because they're not legal Antlr nonterminal names. But, maybe "children", and whatever else is being used as a symbol name but that's a method or field name in the generated code. And, move [the actual "badWords" list](https://github.com/antlr/antlr4/issues/1070#issuecomment-266243640) to this mapping.

**Comment by KvanTTT** [comment]
> Goodness, why wasn't this mapping added for each of the other targets?

Because other targets don't support keyword escaping.

> Symbol rewrite in the generated code was suggested in 2015 but nothing came of it.

Yes, it's a good idea. Just add `_` suffix to conflicting identifiers at the generation step.

**Comment by kaby76** [comment]
> > Goodness, why wasn't this mapping added for each of the other targets?
> 
> Because other targets don't support keyword escaping.

Good grief. We didn't add the code to the .stg's (and rip out "badWords" elsewhere) because it's not already added.

This is really an easy change. Will give a try.

**Comment by kaby76** [comment]
StringTemplate is delightful code! I just wish it was possible to pass the templates for code gen into the Antlr tool via command line, so I could write custom parsers, and new targets. As a prototype, I emptied out the "badWords" list for the Java target, then refactored it and added it to Java.stg. I also needed to change a few places where names need to be mapped to altered form. I tried it out on a grammar with "null" and it works beautifully. Attached is the Java.stg code.
[Java.txt](https://github.com/antlr/antlr4/files/6620323/Java.txt). It should be easy to change the other templates and take care of symbol conflicts once and for all.

**Comment by KvanTTT** [comment]
It would be great!

>  Attached is the Java.stg code.

I recommend using [GitHub Gist](https://gist.github.com/) or another code-sharing service because they provide code highlighting, editing, and other useful features.

**Comment by KvanTTT** [comment]
@kaby76 I'm not sure that placing of such keywords into `.stg` code is a very good idea:

1. It's redundant because all dictionary values corresponds to their keys plus suffix (`@` for C#) or prefix (`_` for others).
2. In theory it's possible that replaced value also conflicts with runtime keywords. Especially it's actual for C languages where `_` using is widespread.

Instead I suggest replacing on ANTLR side. We can have `prefix` or `suffix` fields for every runtime. For instance, C# will have empty suffix but `@` prefix. Swift will have both suffix and prefix (`). Other runtimes will have `_` suffix.

I'm going to fix it in my ANTLR fork together with other tasks that I consider important especially for grammars (first of all, introducing of case insensitive option).

**Comment by kaby76** [comment]
I agree, just rename the generated symbols so there won't be a conflict
ever. Ideally I'd like to pass the templates to the tool to provide even
more flexibility.

**Comment by NickStrupat** [comment]
Name collisions could be resolved by placing the `new` modifier on the generated lexer/parser class members which collide with the library member.

The `EMPTY` token member can be retrieved via `<LexerName>Lexer.EMPTY` and the original `ParserRuleContext.EMPTY` member can be retrieved by static cast of your instance to `ParserRuleContext`.

**Comment by KvanTTT** [comment]
We are working on the problem of conflicting identifiers.

**Comment by NickStrupat** [comment]
@KvanTTT you're awesome! Wondering if this issue I found might be related or of interest...

https://github.com/antlr/antlr4/issues/3503
